drivers/bus/isapnp/hardware.c

```diff
@@ -1570,64 +1518,62 @@ IsaHwConfigureDevice(
         {
             case CmResourceTypePort:
             {
-                if (++NumberOfIo > RTL_NUMBER_OF(LogicalDevice->Io))
+                if (NumberOfIo >= RTL_NUMBER_OF(LogicalDevice->Io))
                     return STATUS_INVALID_PARAMETER_1;
 
-                if (!FindIoDescriptor(LogicalDevice,
-                                      0,
-                                      Descriptor->u.Port.Start.LowPart,
-                                      Descriptor->u.Port.Start.LowPart +
-                                      Descriptor->u.Port.Length - 1,
-                                      NULL,
-                                      NULL))
-                {
-                    return STATUS_RESOURCE_DATA_NOT_FOUND;
-                }
-
+                IsaProgramIoDecoder(Descriptor, NumberOfIo++);
                 break;
             }
 
             case CmResourceTypeInterrupt:
             {
-                if (++NumberOfIrq > RTL_NUMBER_OF(LogicalDevice->Irq))
+                if (NumberOfIrq >= RTL_NUMBER_OF(LogicalDevice->Irq))
                     return STATUS_INVALID_PARAMETER_2;
 
-                if (!FindIrqDescriptor(LogicalDevice, Descriptor->u.Interrupt.Level))
-                    return STATUS_RESOURCE_DATA_NOT_FOUND;
-
+                IsaProgramIrqSelect(Descriptor, NumberOfIrq++);
                 break;
             }
 
             case CmResourceTypeDma:
             {
-                if (++NumberOfDma > RTL_NUMBER_OF(LogicalDevice->Dma))
+                if (NumberOfDma >= RTL_NUMBER_OF(LogicalDevice->Dma))
                     return STATUS_INVALID_PARAMETER_3;
 
-                if (!FindDmaDescriptor(LogicalDevice, Descriptor->u.Dma.Channel))
-                    return STATUS_RESOURCE_DATA_NOT_FOUND;
-
+                IsaProgramDmaSelect(Descriptor, NumberOfDma++);
                 break;
             }
 
             case CmResourceTypeMemory:
             {
-                BOOLEAN Memory32;
+                BOOLEAN IsMemory32;
+                UCHAR Index, Information;
+                NTSTATUS Status;
 
-                if (++NumberOfMemory > RTL_NUMBER_OF(LogicalDevice->MemRange))
+                if ((NumberOfMemory + NumberOfMemory32) >= RTL_NUMBER_OF(LogicalDevice->MemRange))
                     return STATUS_INVALID_PARAMETER_4;
 
                 if (!FindMemoryDescriptor(LogicalDevice,
```

Comment:
Is it intended that now, the handling of `CmResourceTypeMemory` is looking "slightly" different from the other resources?
What I mean is this: `CmResourceTypeMemory` still has the `FillXXXXDescriptor()` call, followed by explicit increment of `NumberOfXXXX`, followed by `IsaProgramXXXXDecoder`, while the other resources now have only a `IsaProgramYYYYDecoder` call.

Reply:
This is due to the fact that the extra information is stored in a PNP ROM. We have to determine how to program the memory control register when configuring the device.

Reply:
It may be of interest to add some explanation along the lines of your answer, as a comment in the source code :thinking:

Reply:
Done.